### ISSUE TITLE
Serializing class with typed Closure property does not work

### ISSUE THREAD

**Issue by davidbyoung** [opis/closure#45]
If we attempt to serialize a class that contains a typed `Closure` property (PHP 7.4), we get a type error.

```php
class Foo
{
    public \Closure $closure;
}

$foo = new Foo();
$foo->closure = function () {
    echo 'hi';
};
$serializedFoo = \Opis\Closure\serialize($foo);
$unserializedFoo = \Opis\Closure\unserialize($serializedFoo);
$unserializedFoo->closure(); // Expect 'hi', get PHP Fatal error:  Uncaught TypeError: Typed property Foo::$closure must be an instance of Closure, Opis\Closure\SerializableClosure used in C:\test\vendor\opis\closure\src\SerializableClosure.php:428
```

This is likely happening because we're overwriting property values with `SerializableClosure` per [this code](https://github.com/opis/closure/blob/master/src/SerializableClosure.php#L428).  However, we cannot do that with typed properties in PHP 7.4.

**Comment by msarca** [comment]
Given the fact that PHP 7.4 is an important update, maybe it's time to release a major version of Opis Closure that will target only PHP7.4 and its subsequent versions. Because it becomes harder and harder to write and maintain code that runs well on both PHP5.4+ & PHP7.  And we'll see then if there is a solution for this kind of issues.

**Comment by davidbyoung** [comment]
I've come across a related issue.  Let's say I have no control over the contents of a closure property, eg

```php
use Closure;
use Opis\Closure\SerializableClosure;

// My class
class Foo
{
    /** @var Closure Purposely not using 7.4 typehint due to issue #45 */
    public $closure;

    public function __sleep(): array
    {
        $this->closure = serialize(new SerializableClosure($this->closure));

        return array_keys(get_object_vars($this));
    }

    public function __wakeup(): void
    {
        /** @var SerializableClosure $wrappedClosure */
        $wrappedClosure = unserialize($this->closure, [Closure::class]);
        $this->closure = $wrappedClosure->getClosure();
    }
}

// A class from userland that I have no control over
class Bar
{
    // Let's assume userland is using typehints
    public Closure $closure;

    public function __construct(Closure $closure)
    {
        $this->closure = $closure;
    }
}

$foo = new Foo();
$bar = new Bar();
$foo->closure = fn () =>($bar->closure)();

// This will not work because of the typehint in Bar
serialize($foo);
```

It looks like `SerializableClosure` recursively attempts to wrap any closure properties so that they can be serialized correctly.  Is there a way to prevent the wrapping of closures for a particular class?  In my real world scenario, I have a DI container that is used inside a closure:

```php
$container = new Container();
// ...
$commandHandlerFactory = fn () => $container->resolve(SomeCommandHandler::class);
file_put_contents('commandFactory.txt', serialize($commandHandlerFactory)); // Throws same error as #45
```

I have explicit logic in my [DI container](https://github.com/aphiria/aphiria/blob/master/src/DependencyInjection/src/Container.php#L49) to not even serialize nor deserialize for performance reasons (common in other DI container libraries, too).  So, wrapping the container's closures isn't really necessary.  Is there a clever workaround for this issue?

**Comment by davidbyoung** [comment]
Hi @msarca.  Do you have any rough idea of when typed properties might be fully supported?

**Comment by msarca** [comment]
Closure wrapping was a good idea when typed properties were not supported in PHP. This feature allowed us to take an arbitrary object and serialize it. But starting with PHP 7.4 typed properties are a fact and, as I said above, a major update of the library is necessary.  But no major update will ever change the fact that we'll never be able to store a `SerializableClosure` object on a `Closure` typed property.  

This practically makes the current wrap/unwrap mechanism to be useless for classes that contains typed properties. As the time passes and people start to adopt PHP 7.4, those classes would be encountered more and more often and the wrap/unwrap mechanism will cause a lot of frustration. That's why the next major version is about finding something that will replace the current mechanism since it can't be fixed. 

Until then a partial solution would be to explicitly wrap closures  at serialization time. 
```php
use Opis\Closure\SerializableClosures;

class Foo
{
    private Closure $property;
    
    public function __construct(Closure $closure)
    {
        $this->property = $closure;
    }
    
    public function __serialize(): array
    {
        return [
            'property' => SerializableClosures::from($this->property),
        ];
    }
    
    public function __unserialize(array $data): void
    {
       $this->property = $data['property']->getClosure();
    }
}
```

**Comment by davidbyoung** [comment]
Yeah, that solution would work for most cases.  Unfortunately, in my particular situation, the class with the typed `Closure` property is composed by a DI container that is being wrapped (same issue as #51).  So, implementing `Serializable` does not prevent `Closure` properties from being wrapped recursively.

**Comment by davidbyoung** [comment]
Is there any update on supporting typed `Closure` properties?  I'd love to help, but considering it would be a near rewrite of this library, I'm afraid I wouldn't be able to tackle it myself.

**Comment by msarca** [comment]
Hi, David. Sorry for the delayed response. Today we've ran a first test using a new approach for serialising closures. It was a success and we are confident to say that probably next week we'll have a public branch of Opis Closure 4.0. This new approach will also fix issue #51 and it will target both PHP 7.4 and PHP 8.0.

**Comment by msarca** [comment]
This should work with Opis Closure 4.x. See #59
